src/network/serialize.rs

```diff
@@ -19,13 +19,100 @@
 //! It also defines (de)serialization routines for many primitives.
 //!
 
+use std::error;
+use std::fmt;
+use std::io;
 use std::io::{Cursor, Read, Write};
 use byteorder::{LittleEndian, WriteBytesExt, ReadBytesExt};
 use hex::encode as hex_encode;
 
+use bitcoin_bech32;
+
 use network::encodable::{ConsensusDecodable, ConsensusEncodable};
+use util::base58;
 use util::hash::Sha256dHash;
-use util;
+
+/// Serialization error
+#[derive(Debug)]
+pub enum Error {
+    /// And I/O error
+    Io(io::Error),
+    /// Base58 encoding error
+    Base58(base58::Error),
+    /// Bech32 encoding error
+    Bech32(bitcoin_bech32::Error),
+    /// Error from the `byteorder` crate
+    ByteOrder(io::Error),
+    /// Network magic was not what we expected
+    BadNetworkMagic(u32, u32),
+    /// Network message was unrecognized
+    BadNetworkMessage(String),
+    /// Parsing error
+    ParseFailed,
+    /// Error propagated from subsystem
+    Detail(String, Box<Error>),
```

Comment:
We should drop the `Box<Error>` from this; the only thing ever packaged in it is `Error::ParseFailed`. In fact I think we should just make `ParseFailed` take a `String`.

Reply:
In fact it should take a `&'static str` rather than a `String`. The only place we dynamically generate a string is in `util::hex_bytes`, which (a) is only used in unit tests, (b) will be replaced alongside the hashes in the medium-term future.

Reply:
I believe that the `error(&mut self, err: String) -> Error` method of `RawDecoder`s also use a dynamic `String`, I do believe however, that with this overall change, we can eliminate that method and directly return `serialize::Error`s, does that sound good?

Reply:
@apoelstra I have replaced those instanced with their own `serialize::Error` vairants... What do we do with the `util::hex_bytes` case? Do I do the same?